### ISSUE TITLE
MEAN Security

### ISSUE THREAD

**Issue by apitts** [meanjs/mean#606]
Has anyone done a deep dive into security yet? I'm wondering whether issues like those detailed below exist or not:
http://blog.websecurify.com/2014/08/hacking-nodejs-and-mongodb.html

**Comment by ilanbiala** [comment]
@apitts would you like to address this issue?

**Comment by apitts** [comment]
Thanks @ilanbiala. I'm investigating it at the moment. Just wanted to check if anyone had previously reviewed. I will certainly submit any PRs if I find issues...but I'm not certain that there are issues....

**Comment by lirantal** [comment]
@apitts did you find anything new?
the specific method of attack wouldn't work on MEAN.JS 0.4.0 branch, mostly because we are using passport library which knows how to handle these security concerns.

**Comment by apitts** [comment]
@lirantal I haven't found any issues yet! Yep, I definitely thought that passportjs should prevent the kind of attacks detailed in the article.

**Comment by jloveland** [comment]
additionally, we should test mean.js with certsimple.com and follow some of the recommendations to get an A+ rating :smiley: 
https://certsimple.com/blog/a-plus-node-js-ssl

**Comment by ilanbiala** [comment]
@jloveland what's their accreditation? Is there any reason we should trust them?

**Comment by jloveland** [comment]
@ilanbiala I heard about ssllabs from listening to Steve Gibson's "Security Now" Podcast a while ago:
https://www.grc.com/sn/sn-396.txt
https://www.grc.com/securitynow.htm

If you search for "test ssl certificate" you will ssllabs at the top and other wikis referencing them.

Either way, you can use other tools like digicert to validate the findings of ssllabs.
https://www.digicert.com/help/

We would need a PR with the recommendations from the certsimple blog I referenced, and someone with a trusted cert from a certificate authority.

In September, we will be able to get free trusted certs from letsencrypt. They also automate the management of the cert on your webserver..
https://letsencrypt.org/

*"The objective of Let’s Encrypt and the ACME protocol is to make it possible to set up an HTTPS server and have it automatically obtain a browser-trusted certificate, without any human intervention. This is accomplished by running a certificate management agent on the web server."*

**Comment by lirantal** [comment]
I'm not sure that adding some default ssl cert which is ultra-secure is a requirement though.
Most people who would choose to work with the project in SSL mode will definitely create their own SSL cert.

**Comment by jloveland** [comment]
@lirantal you are right, I did not mean to suggest committing a ssl cert. I mean't that we would need someone with a trusted cert from a certificate authority to test this for us. 

I do think it would be beneficial to add some of the recommendations to the nodejs server if running in *secure* mode:
```
var server = https.createServer({
    key: privateKey,
    cert: certificate,
    ca: certificateAuthority,
    ciphers: [
      "ECDHE-RSA-AES256-SHA384",
      "DHE-RSA-AES256-SHA384",
      "ECDHE-RSA-AES256-SHA256",
      "DHE-RSA-AES256-SHA256",
      "ECDHE-RSA-AES128-SHA256",
      "DHE-RSA-AES128-SHA256",
      "HIGH",
      "!aNULL",
      "!eNULL",
      "!EXPORT",
      "!DES",
      "!RC4",
      "!MD5",
      "!PSK",
      "!SRP",
      "!CAMELLIA"
    ].join(':'),
    honorCipherOrder: true
}, app);
```

```
var helmet = require('helmet');

var ONE_YEAR = 31536000000;
app.use(helmet.hsts({
    maxAge: ONE_YEAR,
    includeSubdomains: true,
    force: true
}));
```

**Comment by lirantal** [comment]
@jloveland adding those cipher options when running in SSL mode means we will need to maintain this list in the future when updates will arrive (new/obsolete ciphers), doesn't it?

**Comment by jloveland** [comment]
@lirantal , yes, ciphers could change as a result of updates to industry standards and security research. The fact is that node still This can be tested periodically by users running ssl and checking their website with https://www.ssllabs.com

There are certain ciphers that should be disabled for known vulnerabilities. This may be viewed by the community as opinionated, but the Nodejs documentation has recommendations and describes each cypher. This could be left up to the developer, but at least we could offer this as a good starting point. For example, RC4 is a well known fallback that should be disabled. They also recommend disabling AES-256 SHA, even though it's enabled by default.

https://nodejs.org/api/tls.html#tls_tls_getciphers
> ***RC4*** is used as a fallback for clients that speak on older version of the TLS protocol. RC4 has in recent years come under suspicion and ***should be considered compromised for anything that is truly sensitive.*** It is speculated that state-level actors possess the ability to break it.

> ***NOTE***: Previous revisions of this section suggested AES256-SHA as an acceptable cipher. Unfortunately, AES256-SHA is a CBC cipher and therefore susceptible to ***BEAST*** attacks. ***Do not use it.***

**Comment by lirantal** [comment]
@ilanbiala I tend to lean towards adding those ciphers to indeed provide a secure environment. What do you think? is that too much for the framework or is aligned with providing a good and secure OOTB nodejs config?

**Comment by lirantal** [comment]
@jloveland how did you compile that specific list of ciphers?
```
ciphers: [
      "ECDHE-RSA-AES256-SHA384",
      "DHE-RSA-AES256-SHA384",
      "ECDHE-RSA-AES256-SHA256",
      "DHE-RSA-AES256-SHA256",
      "ECDHE-RSA-AES128-SHA256",
      "DHE-RSA-AES128-SHA256",
      "HIGH",
      "!aNULL",
      "!eNULL",
      "!EXPORT",
      "!DES",
      "!RC4",
      "!MD5",
      "!PSK",
      "!SRP",
      "!CAMELLIA"
    ].join(':'),
```

**Comment by lirantal** [comment]
@jloveland also for helmet's hsts, `ONE_YEAR` is just as an example or do you have a reason for it being very long? seems like 18 weeks approved by Google, is good enough. WDYT?

**Comment by jloveland** [comment]
I got this list from https://certsimple.com/blog/a-plus-node-js-ssl. Looking at the article update, you can see the mention that they updated the default ciphers in the next version of nodejs.
>Update May 18 2015: [CertSimple has added newer, even better ciphers into node.js itself, which will appear in the next release of node](https://github.com/nodejs/node/commit/5755fc099f883293530406c423bda47414834057). If you're using node 0.12, [grab the cipher list from the latest node source](https://github.com/nodejs/node/blob/master/lib/tls.js) which contains our changes, then add HSTS support per this article.

Looking at the new nodejs source, this is the default (***note*** they added GCM ciphers for Chrome's 'modern cryptography' setting):
```
exports.DEFAULT_CIPHERS = [
  'ECDHE-RSA-AES128-GCM-SHA256',
  'ECDHE-ECDSA-AES128-GCM-SHA256',
  'ECDHE-RSA-AES256-GCM-SHA384',
  'ECDHE-ECDSA-AES256-GCM-SHA384',
  'DHE-RSA-AES128-GCM-SHA256',
  'ECDHE-RSA-AES128-SHA256',
  'DHE-RSA-AES128-SHA256',
  'ECDHE-RSA-AES256-SHA384',
  'DHE-RSA-AES256-SHA384',
  'ECDHE-RSA-AES256-SHA256',
  'DHE-RSA-AES256-SHA256',
  'HIGH',
  '!aNULL',
  '!eNULL',
  '!EXPORT',
  '!DES',
  '!RC4',
  '!MD5',
  '!PSK',
  '!SRP',
  '!CAMELLIA'
].join(':');
```

Until we update to the latest version of Nodejs, this list seems like the best. We could use another person to review to make sure I am interpreting this correctly.

**Comment by jloveland** [comment]
for helmet's hsts, I have read on a couple articles that we should set it to 6 months.
https://blog.cloudflare.com/enforce-web-policy-with-hypertext-strict-transport-security-hsts/
>Max Age (max-age): This is essentially a "time to live" field for the HSTS header. We recommend 6 months in order to earn an A+ rating from Qualys SSL Labs. Web browsers will cache and enforce HSTS policy for the duration of this value. A value of "0" will disable HSTS.

**Comment by ilanbiala** [comment]
@lirantal I don't have an issue with adding more SSL support, I just want to make sure we have a reputable source for this information, because I haven't heard of SSL Labs and CertSimple.

**Comment by codydaig** [comment]
@ilanbiala I believe this was just merged in #765
